### DISCUSSION TITLE
Zero-padding numbers in interpolation

### DISCUSSION THREAD

**Discussion by madhadron** [cue-lang/cue#1452] [Q&A] [answered]
I have a list of integers that I am going to be interpolating into a list of strings, e.g.,

```
numbers: [1, 6, 24, 42]
labels: [for n in numbers { "endpoint\( n )" }]
```

but I would like the number to always be two digits, e.g., endpoint06, endpoint24. I can do something a little gross like

```
numbers: [1, 6, 24, 42]
labels: [for n in numbers if n < 10 { "endpoint0\(n)" }] +
    [for n in numbers if n >= 10 { "endpoint\(n)" }]
```

or something roundabout like

```
t: {
    _n: number & >=0 & <10,
    endpointUri: "endpoint0\(_n)"
} | {
    _n: number & >=10 & <100,
    endpointUri: "endpoint\(_n)"
}

numbers: [1, 6, 24, 42]
labels: [for n in numbers { t & {_n: n}}]
```

but is there a simpler way to zero-pad a number?

**Comment by xinau** [comment, marked as answer]
@madhadron I think there is currently no easier way as there doesn't seem to exist a `fmt.Printf` in CUE's [stdlib][stdlib]. If you prefer you could also do it more inline, but I'm not sure if this is an improvement.

```cue
numbers: [1, 6, 24, 42]

// solution 1
#padding: {
	number:   <10
	endpoint: "0\(number)"
} | {
	number:   >=10
	endpoint: "\(number)"
}

labels: [ for n in numbers {"endpoint\((#padding & {number: n}).endpoint)"}]

// solution 2
labels: [ for n in numbers {"endpoint\({if n < 10 {"0"}, if n >= 10 {""}})\(n)"}]
```

I think this might be a good case to provide `printf` functionality in CUE's [stdlib][stdlib].

[stdlib]: https://pkg.go.dev/cuelang.org/go@v0.4.0/pkg

**Reply by mpvl** [reply]
I would be fine having a `fmt`-like package in CUE in lieu of something better.

**Reply by madhadron** [reply]
Yeah, sprintf is gross, but it works and everyone knows it. That seems like a reasonable step.

**Reply by myitcv** [reply]
Raised https://github.com/cue-lang/cue/issues/1463 to capture this

**Comment by jsok** [comment]
Not sure if simpler, but you can use `printf` via `text/template`, e.g. https://cuelang.org/play/?id=Rp23O2xDZnM#cue@export@cue

**Reply by madhadron** [reply]
That's a solid workaround for now, thanks!
